src/html/Quoting.cc

```diff
@@ -9,81 +9,74 @@
 #include "squid.h"
 #include "html/Quoting.h"
 
-/*
- *  HTML defines these characters as special entities that should be quoted.
- */
-static struct {
-    unsigned char code;
-    const char *quote;
-} htmlstandardentities[] =
+#include <array>
+#include <cstring>
+
+class HtmlSpecialCharacters {
+private:
+    std::array<const char *, 256> escapeMap = {};
```

Comment:
Can we use SBuf to store escape sequences, removing the need for adding this (_currently buggy_) new class (and some associated hacks like maxEscapeLength with it)?

```suggestion
    std::array<SBuf, 256> escapeMap = {};
```

Reply:
This would have the effect of pulling libsbuf, libbase, mempools into everythin in $(top_builddir)/tools , via libhtml.la , so.. no. I could use (for now) std::string

Reply:
> This would have the effect of pulling libsbuf, libbase, mempools into everythin in $(top_builddir)/tools

You are right. It looks like an unfortunate typo in [previous](https://github.com/squid-cache/squid/pull/1513#issuecomment-1759609500) [analysis](https://github.com/squid-cache/squid/pull/1513#issuecomment-1759663076) resulted in this tools/cachemgr problem unaddressed by the recent official commit 80cba6d. Sigh. FWIW, I would (still) enthusiastically support removing cachemgr from this repository.

> I could use (for now) std::string

We should not use std::string in new src/ code (I can detail the reasons if needed, but we have discussed this before). There are some narrow exceptions to that rule, but src/html/Quoting.cc does not really qualify. The correct solution, IMO, is to remove cachemgr. We could move cachemgr into src/ as a workaround, but that would only prolong that tool's agony.

We could grant Quoting.cc a temporary exception for std::string use to save time required to fix HtmlSpecialCharacters class, but such an exception may encourage even more wasteful attempts to sprinkle src/ with std::string. Fortunately, that exception is not necessary to solve the problem this change request is focusing on.

I recommend a different solution:

1. Remove HtmlSpecialCharacters class. This way you do not have to fix it (or even discuss how it is broken).
2. "Leak" std::array items memory. These items are allocated only once in program lifetime. It is acceptable to never destroy them (Squid has lots of similar "leaks" and even _recommends_ them as the best solution for the "No new globals" rule which is very applicable to this array).

After the above changes, the beginning of html_quote() may look like this:

```C++
static const auto escapeSequences = MakeEscapeSequences();
```

... where MakeEscapeSequences() is a `static` function that allocates (and returns a pointer to) std::array<const char *, 256>. Try using `auto` for MakeEscapeSequences() return type so that you do not have to duplicate or name that array type -- it should only be spelled out once.

Reply:
Done and working, with a slight difference: instead of a pointer, I'm returning a reference to a static variable in MakeEscapeSequences(). Is it an acceptable change? What are the pros/cons, if any?

Reply:
> instead of a pointer, I'm returning a reference to a static variable in MakeEscapeSequences(). Is it an acceptable change?

No, returning a reference to a disappearing (static) variable is not acceptable. See the documentation for the "[No new globals](https://wiki.squid-cache.org/DeveloperResources/SquidCodingGuidelines#rule-no-new-globals)" rule for explanations, especially the last paragraph.

Reply:
Hm.. "adjusted to return a constant reference" in the "no new globals" spec is confusing :\

I'm not particularly keen to mix pointers with operator[]. Since it's done only once, what about copying? I'll use pointers if you insist

Reply:
I insist (and can explain why if needed). I can adjust this myself if you prefer.

Reply:
> "adjusted to return a constant reference" in the "no new globals" spec is confusing :\

In what way? The sketched function returns a constant reference. It can be adjusted to return a constant reference (instead of a non-constant one).

BTW, you _can_ implement this using the sketch from "no new globals". Doing so will eliminate pointers to the array from html_quote().